Slider: Enforce numeric constraints and styling within the text input

drew08t opened the pull request:
What this PR does / why we need it:
Improved state management is needed when using one instance of slider across multiple layers, for example in geomap.

- [x] Add NumberInput to core slider
- [x] Improve state management
- [x] Improve styling for NumberInput adjacent to slider
- [x] Simplify

drew08t committed:
Add NumberInput to core slider

ryantxu commented on public/app/core/components/OptionsUI/slider.tsx:
```diff
@@ -1,26 +1,85 @@
+          marks={marks}
+          included={included}
+        />
+        {/* Uses text input so that the number spinners are not shown */}
```
except on hover... and despite all my efforts, will take a change in grafana/ui to fix that

ryantxu commented on public/app/core/components/OptionsUI/slider.tsx:
```diff
@@ -1,26 +1,85 @@
+          included={included}
+        />
+        {/* Uses text input so that the number spinners are not shown */}
+        <NumberInput value={sliderValue} onChange={onSliderInputChange} max={settings?.max} min={settings?.min} />
```
```suggestion
        <NumberInput value={sliderValue} onChange={onSliderInputChange} max={max} min={min} step={step} />
```
the number values should match the slider -- if the slider is auto clamped to 0-100, so should the text

drew08t committed:
Change opacity interpretation

To be consistent with other layers, and CSS, opacity of 0 is fully
transparent and an opacity of 1 is fully opaque.

drew08t committed:
Add state management for slider.

drew08t committed:
Ensure number input step matches slider

drew08t committed:
Style input width based on expected digits

drew08t committed:
Consolidate react import

drew08t commented:
Slider + NumberInput styling:
![Jun-16-2022 20-05-39](https://user-images.githubusercontent.com/60050885/174216868-7a0ce331-4717-4731-900e-971e192d68e2.gif)

drew08t commented on public/app/core/components/OptionsUI/slider.tsx:
```diff
@@ -85,8 +106,36 @@ export const SliderValueEditor: React.FC<FieldConfigEditorProps<number, SliderFi
+      min-width: ${width}px;
+      width: 100%;
+    `,
+    numberInputHidden: css`
```
Need to remove this.

ryantxu commented on public/app/core/components/OptionsUI/slider.tsx:
```diff
@@ -1,26 +1,141 @@
 };
+
+// Calculate width of string with given font
+function getTextWidth(text: string, font: string): number | null {
```
is `measureText` not good enough?

drew08t replied:
I am using that below on line 123, but it requires context to return the TextMetrics object.

drew08t committed:
Clean up organization

ryantxu commented on public/app/core/components/OptionsUI/slider.tsx:
```diff
@@ -1,26 +1,143 @@
+import { FieldConfigEditorProps, GrafanaTheme2, SliderFieldConfigSettings } from '@grafana/data';
+import { useTheme2 } from '@grafana/ui';
+
+import { getStyles } from '../../../../../packages/grafana-ui/src/components/Slider/styles';
```
```suggestion
import { getStyles } from '@grafana/ui/src/components/Slider/styles';
```

ryantxu commented:
Is this ready for review yet?

drew08t committed:
Improve styles import path

ryantxu commented:
Here is some kinda funky behavior... but not really sure how to best improve it:
![2022-07-06_14-26-14 (1)](https://user-images.githubusercontent.com/705951/177646549-e0a4bbbc-57e3-4e34-af73-66180e35f8fb.gif)

ryantxu commented:
Since this has "add-to-changelog" we should make sure the title is meaningful :) maybe:

`Slider: Enforce numeric constraints within the text input`. or something

drew08t commented:
> Here is some kinda funky behavior... but not really sure how to best improve it: ![2022-07-06_14-26-14 (1)](https://user-images.githubusercontent.com/705951/177646549-e0a4bbbc-57e3-4e34-af73-66180e35f8fb.gif) [ ](https://user-images.githubusercontent.com/705951/177646549-e0a4bbbc-57e3-4e34-af73-66180e35f8fb.gif) ![2022-07-06_14-26-14 (1)](https://user-images.githubusercontent.com/705951/177646549-e0a4bbbc-57e3-4e34-af73-66180e35f8fb.gif) [ ](https://user-images.githubusercontent.com/705951/177646549-e0a4bbbc-57e3-4e34-af73-66180e35f8fb.gif)

The warning that pops up and impacts the nearby element styles?

drew08t commented:
Looks like the styling behavior could be addressed here in Field.tsx:
![image](https://user-images.githubusercontent.com/60050885/177658612-3cc6d372-3403-4ae5-8c64-8fd8f0a2873d.png)

drew08t commented:
@ryantxu If you like this styling instead, I can commit it. Best I could do without going deeper and modifying form Field.

![Jul-06-2022 18-34-05](https://user-images.githubusercontent.com/60050885/177671526-b7fe84b5-9a22-4c25-abb6-3c6e4b9eb74d.gif)

drew08t committed:
Improve styling for number input validation error

drew08t committed:
Revert change to styles as no longer needed
